### DISCUSSION TITLE
USB3 connectivity

### DISCUSSION THREAD

**Discussion by jdm0830** [debauchee/barrier#1558] [Ideas]
hello. I wish to make a more secure channel by connecting USB 3.0 over my two macs. then having barrier talk over that. I tried creating ethernet connections out of my USBs, but I failed with what knowledge I had. Would help not only privacy but performance. I always have a bit of spotty connectivity with my wireless everywhere I go between my two macs.

**Comment by jdm0830** [comment]
can I pay to have the USB feature enabled? my biggest beef is that my mouse is not smooth. im guessing it cause its soooo slow on wifi

**Comment by farhansolodev** [comment]
bumping this

**Comment by mandusm** [comment]
I have a very similar request to this. I would love to see Barrier support some kind of USB HID Gadget. Something like a Raspberry Pi Zero or ESP32 that I can plug into any device, connect that HID gadget to my Host computer, and then control it through Barrier.

This will allow me to use my iPad Pro like an extended display using barrier without the need for any software to be installed.

**Reply by windoze** [reply]
I know this is an old thread but I just made the thing you're looking for, Barrier client on ESP32S3.
You can check it out at https://github.com/windoze/esparrier

**Comment by mitchcapper** [comment]
@jdm0830 while you could create networks over usb3 to usb3 using just the right cable, you could also just use actual usb network adapters with better flexibility (and much longer lengths). You would still get similar security using a private net.

**Comment by windoze** [comment]
USB is a Host/Device architecture, there is exact one host and multiple devices.
Most PC/Mac can only act as host so you cannot use USB cable to connect 2 PCs.

In your case you can use two USB3/NIC adapters and connect 2 PCs with network cable though, of course you need to configure the IP address, net masks, etc.

**Reply by mitchcapper** [reply]
@windoze previously yes but with usb3 it is possible in the spec and some products do this: https://www.amazon.com/Plugable-Transfer-Compatible-Computer-Migration/dp/B01B6X8QP0

but def not the easiest or best way to go.   Technically you probably don't need to configure networks manually, windows at least does implement APIPA.

**Reply by windoze** [reply]
This cable is not a 'cable' but some kind of transmitter, which acts as a device on both sides and exchanges vendor defined data frames. With app/driver installed on both sides, such a device can do many things.
OTHO Thunderbolt does allow you to connect 2 computers, or even more with daisy-chain, but it's a totally different thing even it shares the same physical connector with USB Type-C.

**Reply by mitchcapper** [reply]
it has usb A ends but its requirements for USB 3 likely use the proper USB 3 protocol and just don't have the power feeds.   Macs it turns out support it pretty easily: https://support.apple.com/en-is/guide/mac-help/mchlp1443/mac
and windows has as well since windows 10: https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-dual-role-driver-stack-architecture and windows IPoverUSB.   I am sure there is some linux setup as well.

**Reply by windoze** [reply]
If you dig deeper into these articles, you'll find that they need "USB on-the-go (OTG)" **hardware** support, I'm not sure about how mac implements the "target mode" but most X86 PCs don't have such hardware, even the OSes do have the support driver included, not only macOs and Windows, these driver software has been in Linux kernel for years,  check out [this article](https://www.kernel.org/doc/html/v5.7/usb/gadget_configfs.html) for more info.
Most mobile phones, Raspberry Pi and many other ARM-based SBCs support USB-OTG and can be used as USB peripherals, and you can find many example about how to use RPi as a keyboard out there.

**Reply by mitchcapper** [reply]
All thunderbolt 4 and newer devices (2020) support pc 2 pc through intels own thunderbolt share: https://www.intel.com/content/www/us/en/download/822291/thunderbolt-share.html  and thunderbolt 3 (2015) can simulate a ip network with host to host.

USB 4 essentially is back compat with thunderbolt 3 meaning all USB 4 (2019) PCs also support thunderbolt network: https://learn.microsoft.com/en-us/windows-hardware/design/component-guidelines/usb4-interdomain-connections

Don't own a mac but apples documentation sounds like all macs support it.

The article talks about using linux at a simulated device gadget which certainly has been around for awhile but that article was from 2013.

Linux has supported thunderbolt networking since kernel 4.15 (2018) as well through thunderbolt-net module (and somehow back to thunderbolt 2 even):

> Thunderbolt technology allows software communication across two hosts connected by a Thunderbolt cable.
> It is possible to tunnel any kind of traffic over Thunderbolt link but currently we only support Apple ThunderboltIP protocol.
> If the other host is running Windows or macOS only thing you need to do is to connect Thunderbolt cable between the two hosts, the thunderbolt-net is loaded automatically. If the other host is also Linux you should load thunderbolt-net manually on one host (it does not matter which one):

https://www.kernel.org/doc/html/v4.15/admin-guide/thunderbolt.html


Finally, finding a copy of the usb 3 spec host to host is in the spec including the cable used  which is basically a normal usb 3 cable sans power:
![image](https://github.com/user-attachments/assets/06e346e8-695e-4031-a12d-7ff4273744ef)

**Reply by windoze** [reply]
USB3 host-to-host connection is mainly used for "debugging", and it's a standalone function which needs to be implemented by the HCI controller separately, I'm not sure if recent xHCI controller does but at least Ivy Bridge doesn't. This type of connection does support data transferring, but AFAIK no existing software allows you to setup network on it, all USB3 host-to-host products I've seen are using bridge chips such as Prolific PL2701.
Thunderbolt is something that exports the internal PCIe bus to an external, serial interface, which of course supports much more usages than USB3.
As for USB4, which is very different from USB3, and much closer to Thunderbolt4, does support host-to-host connection, and allow you to setup USB4NET without additional hardware, and most recent OSes already support it.
